### ISSUE TITLE
panic error running last version of tendermint docker

### ISSUE THREAD

**Issue by ellguso** [tendermint/tendermint#1398]
BUG REPORT

version: 0.17.1-6f995699

uname -a: Linux ellguso-pc 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
Installed using go get

When running the docker version of tendermint i get the error:

```sh
panic: Panicked on a Sanity Check: Could not create directory /tendermint/config. mkdir /tendermint/config: permission denied

goroutine 1 [running]:
github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/common.PanicSanity(0xb8a740, 0xc420952f70)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/common/errors.go:27 +0xf3
github.com/tendermint/tendermint/config.EnsureRoot(0xc420016c68, 0xb)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/config/toml.go:30 +0x366
github.com/tendermint/tendermint/cmd/tendermint/commands.ParseConfig(0x12c6320, 0xcbf7c1, 0x7)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/cmd/tendermint/commands/root.go:37 +0xfc
github.com/tendermint/tendermint/cmd/tendermint/commands.glob..func3(0xc420098480, 0xc4208e0500, 0x0, 0x4, 0x0, 0x0)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/cmd/tendermint/commands/root.go:49 +0x84
github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli.concatCobraCmdFuncs.func1(0xc420098480, 0xc4208e0500, 0x0, 0x4, 0x0, 0x0)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli/setup.go:115 +0x80
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).execute(0xc420098480, 0xc4208e04c0, 0x4, 0x4, 0xc420098480, 0xc4208e04c0)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:677 +0x58b
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x12c6540, 0xc42091fe18, 0x7eb9e1, 0xbf6d40)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:783 +0x30e
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).Execute(0x12c6540, 0x7eb297, 0xc4208d83c0)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli.Executor.Execute(0x12c6540, 0xe172a0, 0x2, 0xc420016b20)
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/tmlibs/cli/setup.go:89 +0x4e
main.main()
	/home/jslave/workspace/02.Fetch.Binary@9/go/src/github.com/tendermint/tendermint/cmd/tendermint/main.go:44 +0x235
```

Download the project, in tendermint/ethermint/tests/test.sh  line 61, remove the -d flag to get the output.

I have the same problem using the lastest version from docker 0.17 but this does not seem to happen using version 0.15 (in docker hub)

**Comment by ebuchman** [comment]
Ethermint has not been updated to latest tendermint (it still requires v0.15.0).

That said this looks like some kind of permission error?

**Comment by t-bast** [comment]
I'm having the same issue with Tendermint 0.17.1 and 0.18.0 (wasn't happening with 0.15.0).

It's not linked to ethermint, simply running _docker run -it --rm -v "$(pwd)/tendermint:/tendermint" tendermint/tendermint:0.18.0 init_ on ubuntu 16.04 fails with that error.

It's linked to Tendermint's new folder structure inside the _tendermint_ folder, namely the _config_ and _data_ folders can't be created.
This is most likely a docker volume permissions issue, because if I create those subfolders before running tendermint init then it works.
Probably not a Tendermint issue at all, we just need to tweak how we run the init from inside docker with a volume mount.

**Comment by t-bast** [comment]
I think the issue is linked to the update to alpine 3.7. Tendermint's dockerfile might need to be updated according to the latest answer on this thread: https://github.com/mhart/alpine-node/issues/48

**Comment by greg-szabo** [comment]
The new feature of tendermint (since somewhere 0.15) is that it's creating a sub-folder "config".
The expectation from docker is that it expects the attached volume (folder) to be writable.
The issue is that SELinux policy on Ubuntu 16.04 (and CentOS7 for that matter) does not allow the docker process to write that directory. I forgot the exact underlying policy that needs to be enabled but docker works around it with the `Z` flag (which tells SELinux to allow that policy).

This should work: (note the extra `Z` flag in the volume definition):

```
docker run -it --rm -v "$(pwd)/tendermint:/tendermint:Z" tendermint/tendermint:0.18.0 init
```

**Comment by melekes** [comment]
Please reopen if https://github.com/tendermint/tendermint/issues/1398#issuecomment-381153042 does not work for you. Thank you

**Comment by t-bast** [comment]
Hey guys,

Thanks for the proposal, but unfortunately this isn't the issue.
I don't have SELinux enabled on my system, only AppArmor.
So the ":Z" and ":z" options unfortunately don't change anything.

I tried troubleshooting this further, but unfortunately I quickly get stuck because of the fact that your official images don't seem to have been generated via a _make build_ in your DOCKER folder...

When I clone tendermint and run: ``cd DOCKER && make build``, the resulting image simply doesn't work (tested both on Ubuntu 16.04 and MacOS).
I get the following error: ``standard_init_linux.go:195: exec user process caused "no such file or directory"`` as if a dynamic dependency is missing on the system.

So I tried picking inside the official tendermint image (tested on 0.17.1 and 0.18.0) and here is what I think happens:

- The Tendermint image initially runs as root, creates the /tendermint folder, creates the _tmuser_ user and gives it ownership of the /tendermint folder inside the container
- Unfortunately, tmuser's UID and GID (inside the container) don't exist on the host (in my case tmuser's uid is 105 and gid is 106)
- When I run docker run -it --rm -v "$(pwd)/tendermint:/tendermint" tendermint/tendermint:0.18.0 init, it fails with the permissions error. If I look at the permissions on the host, **the tendermint folder is now owned by root**. I suspect (I'm not sure though) that this is because the uid:gid set inside the container don't exist on the host.
- At that point it's a mess because inside the container, the permission has been reset to root on the tendermint folder, so when the image executes (apparently with the tmuser user) it can't do anything to the /tendermint directory, hence the issue.

Do you have an idea on how that could be fixed? I'm not at all a docker expert so any advice would be helpful :)

**Comment by melekes** [comment]
```
» docker run -it --rm -v "/tmp:/tendermint" tendermint/tendermint:0.18.0 init                                                                                  1 ↵
I[04-16|13:15:08.311] Generated private validator                  module=main path=/tendermint/config/priv_validator.json
I[04-16|13:15:08.315] Generated genesis file                       module=main path=/tendermint/config/genesis.json
» docker run -it --rm -v "/tmp:/tendermint" tendermint/tendermint:0.18.0 node --proxy_app=kvstore
I[04-16|13:16:19.731] Starting multiAppConn                        module=proxy impl=multiAppConn
I[04-16|13:16:19.731] Starting localClient                         module=abci-client connection=query impl=localClient
I[04-16|13:16:19.731] Starting localClient                         module=abci-client connection=mempool impl=localClient
```

`0.18` works for me

Docker version 18.04.0-ce, build 3d479c0

**Comment by t-bast** [comment]
It works for me as well when mounting to /tmp, but not mounting to a user folder (/home/bast/scratch:/tendermint). Do you repro in that case?

**Comment by melekes** [comment]
no, hmm

**Comment by greg-szabo** [comment]
If the command doesn't run, please share the exact command-line that you executed.

Here's some more insight:

1. Please check if SELinux is enabled on your system. As far as I remember, it is enabled by default on Ubuntu.
AppArmor might also have its own opinion about Docker, you'll have to search for that (I don't use it).

```
[greg@ip-10-0-1-132 ~]$ sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   enforcing
Mode from config file:          enforcing
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Max kernel policy version:      28
```

2. You can "look into" the container by running the `/bin/sh` command as entrypoint:
```
gregbook:~ greg$ docker run --rm -it --entrypoint /bin/sh tendermint/tendermint
~ $ /usr/bin/tendermint version
0.18.0-4930b61a
~ $ id
uid=100(tmuser) gid=1000(tmuser) groups=1000(tmuser)
~ $ ls -la /tendermint
total 12
drwxr-sr-x    2 tmuser   tmuser        4096 Apr 16 21:51 .
drwxr-xr-x    1 root     root          4096 Apr 16 21:51 ..
-rw-------    1 tmuser   tmuser          50 Apr 16 21:52 .ash_history
```

This will show you that the user inside the docker container is running with uid=100 and gid=1000.

3. The latest image was rebuilt for alpine from this Dockerfile: https://github.com/tendermint/tendermint/blob/greg/devopstools/DOCKER/Dockerfile
We're moving back to alpine, only 0.17.1 remains on an Ubuntu image. (The Dockerfile is under review for the develop branch and will be merged to master after that.)

4. This is the most relevant to your case: Docker does not do UID/GID translation between the host and the guest by default. (I've seen one exception.)

In this case, within docker you run as tmuser.tmuser, uid=100, gid=1000. This does not make any difference on the host system, you only have access to anything that has uid=100, gid=1000 or similar. Note that in linux, username and group name are not identifiers, only uid and gid are. So you can have a file that has a uid that does not belong to a user in `/etc/passwd`. This means that:
1. If the folder you share with docker does not allow uid=100,gid=1000 to write to it, then you will not be able to write to that folder within docker.
2. If the host system does not have uid=100 or gid=1000, the files will still be created with these permissions and they will show up as numbers in `ls -la`.
3. If the host user does not have permission to read/write a file owner by uid=100,gid=1000, the files created within docker will not be readable by the host user.

This looks messed up, but keep in mind that docker was not really created for users to run one-off containers. It is a container system that should be run as `root` on servers that have no users logged on. (Think of an automated warehouse.) By default, the docker daemon runs as root and the`docker run` commands run as root. Within the docker container your process runs as root too. Hence all files created are with UID=0, GID=0 and even if they aren't, the host root will still be able to access them.

Our image contains a non-privileged user, because if it didn't, it would create all files with uid=0,gid=0 (root.root), hence allowing all kinds of interesting security and annoyance issues.

The only exception I've found is the OSX docker service and application. This one, in most cases somehow (magic?) translated the file permissions within docker to the user running on my host system. Most probably the docker service is built differently on OSX (I didn't need to create a docker group and add my user to it). It's completely possible that you can set up this behaviour on Linux too, but I'm not aware of it. Try to look for other ways to set up docker on linux if this doesn't fit your needs. You can also make sure you match the docker 100/1000 user on your host with the one in docker.

With these in mind, I hope I gave enough ammunition to get a fix for your issue.

**Comment by t-bast** [comment]
Thanks Greg for the detailed write-up.
Now that I see the Dockerfile used from your branch it makes a lot more sense.
Strangely enough, when I use alpine as a base image for tendermint I don't repro the issue (but it can just be that the uid:gid pair ends up being more friendly with what exists on my host).
And you're totally right that I can just rebuild a custom tendermint image for my needs, so I do have everything I need to move forward, no need to waste more of your time!

_The only exception I've found is the OSX docker service and application._: I've noticed that as well, it feels like pure magic on OSX I don't know how they got that working...

Small note: SELinux isn't enabled on my machine, it looks like Ubuntu uses AppArmor by default instead. I'll experiment with SELinux to see if it's a better fit than AppArmor for my day to day dev.

I can't close the issue because I wasn't the creator, but feel free to close it. Thanks for your help!
